src/main/java/com/azure/cosmos/samples/distributedbulk/DocumentBulkExecutor.java

```diff
@@ -0,0 +1,117 @@
+package com.azure.cosmos.samples.distributedbulk;
+
+import com.azure.cosmos.CosmosAsyncContainer;
+import com.azure.cosmos.models.CosmosBulkOperations;
+import com.azure.cosmos.models.CosmosItemOperation;
+import com.azure.cosmos.models.PartitionKey;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+public class DocumentBulkExecutor<T> {
+    private final static Logger logger = LoggerFactory.getLogger(DocumentBulkExecutor.class);
+    private final CosmosAsyncContainer cosmosAsyncContainer;
+    private final Function<T, String> idExtractor;
+
+    public DocumentBulkExecutor(
+        CosmosAsyncContainer cosmosAsyncContainer,
+        Function<T, String> idExtractor) {
+
+        Objects.requireNonNull(cosmosAsyncContainer, "Argument 'cosmosAsyncContainer' must not be null.");
+        Objects.requireNonNull(idExtractor, "Argument 'idExtractor' must not be null.");
+        this.cosmosAsyncContainer = cosmosAsyncContainer;
+        this.idExtractor = idExtractor;
+    }
+
+    /*
+    public BulkDeleteResponse deleteAll(List<Pair<String, String>> pkIdPairsToDelete) {
+
+    }
+
+    public BulkImportResponse importAll(Collection<?> documents) {
+    }*/
+
+    /**
+     * Upserts all documents in bulk mode
+     * @param documents a collection of documents for a logical batch
+     * @return A BulkImportResponse with aggregated diagnostics and eventual failures.
+     */
+    public BulkImportResponse upsertAll(Stream<T> documents) {
+        return this.upsertAll(
+            documents,
+            new DocumentBulkExecutorOperationStatus());
+    }
+
+    /**
+     * Upserts all documents in bulk mode
+     * @param documents a collection of documents for a logical batch
+     * @param status can be used to track the status of the bulk ingestion for this batch at real-time
+     * @return A BulkImportResponse with aggregated diagnostics and eventual failures.
+     */
+    public BulkImportResponse upsertAll(
+        Stream<T> documents,
+        DocumentBulkExecutorOperationStatus status) {
+
+        Objects.requireNonNull(documents, "Argument 'documents' must not be null.");
+        Objects.requireNonNull(status, "Argument 'status' must not be null.");
+
+        try {
+            Stream<CosmosItemOperation> docsToUpsert = documents
```

Comment:
Looks like there are at least 3 places where you apply effective throttling, could you correct my understanding?:

- When reading from blob storage, you place a limit on record count - for client memory.
- When scheduling writes in `BulkWriter`, you use a semaphore - for client memory.
- Internal `CosmosBulkExecutor` (batching per I/O call) - for RU.

Reply:
Correct - BulkWriter should be usable independent of DocumentBulkExecutor. The reading form storage in batches and passing them to DocumentBulKExecutor is how V2 customers used BulKExecutor - at that point the BulkWriter's semaphore shouldn't be needed strictly speaking. But to allow usage of BulkWriter independently - on Flux of unbounded incoming stream it is required - so, keeing it here like in Spark